File: promise/src/test/java/com/voxeet/testpromise/resolve/PromiseTestWithSum.java
package com.voxeet.testpromise.resolve;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseExec;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class PromiseTestWithSum {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final int[] final_result = {0};

        System.out.println("executing test");
        execute()
                .then((PromiseExec<Integer, Integer>) (result, solver) -> solver.resolve(10 + result))
                .then((PromiseExec<Integer, Integer>) (result, solver) -> {
                    final_result[0] = 10 + result;
                    latch.countDown();
                })
                .error(error -> {
                    System.out.println("error catched");
                    final_result[0] = 0;
                    error.printStackTrace();
                    latch.countDown();
                });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (final_result[0] != 30) {
            throw new IllegalStateException("Expected 30... having " + final_result[0]);
        } else {
            System.out.println("having result " + final_result[0]);
        }
    }


    @Test
    public void testSoloResolve() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final int[] final_result = {0};

        System.out.println("executing test");
        execute().then((PromiseExec<Integer, Integer>) (result, solver) -> {
            System.out.println("---------");
            final_result[0] = 10 + result;
            latch.countDown();
        }).error(error -> {
            System.out.println("error catched");
            final_result[0] = 0;
            error.printStackTrace();
            latch.countDown();
        });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (final_result[0] != 20) {
            throw new IllegalStateException("Expected 20... having " + final_result[0]);
        } else {
            System.out.println("having result " + final_result[0]);
        }
    }

    private Promise<Integer> execute() {
        return new Promise<>(solver -> solver.resolve(10));
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/resolve/PromiseTestWithSum.kt
package com.voxeet.testpromise.resolve

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.solve.ThenVoid
import com.voxeet.testpromise.mockedhandler
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class PromiseTestWithSum {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
    }

    @Test
    fun test() {
        val latch = CountDownLatch(1)
        var result = 0

        Promise.resolve(10).then { res: Int ->
            10 + res
        }.then(ThenVoid {
            result = 10 + it
            latch.countDown()
        }).error { error: Throwable ->
            println("error catched")
            result = 0
            error.printStackTrace()
            latch.countDown()
        }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(result == 30) { "Expected 30... having $result" }
        println("having result $result")
    }

    @Test
    @Throws(InterruptedException::class)
    fun testSoloResolve() {
        val latch = CountDownLatch(1)
        var result = 0

        Promise.resolve(10).then {
            println("---------")
            result = 10 + it
            latch.countDown()
        }.error { error: Throwable ->
            println("error catched")
            result = 0
            error.printStackTrace()
            latch.countDown()
        }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(result == 20) { "Expected 20... having $result" }
        println("having result $result")
    }
}
